controller/uninstall_controller.go

```diff
@@ -0,0 +1,437 @@
+		force:                     force,
+		firstRun:                  true,
+		ds:                        ds,
+		stopCh:                    stopCh,
+		volumeInformerSynced:      volumeInformer.Informer().HasSynced,
+		engineInformerSynced:      engineInformer.Informer().HasSynced,
+		replicaInformerSynced:     replicaInformer.Informer().HasSynced,
+		engineImageInformerSynced: engineImageInformer.Informer().HasSynced,
+		nodeInformerSynced:        nodeInformer.Informer().HasSynced,
+		daemonSetInformerSynced:   daemonSetInformer.Informer().HasSynced,
+		queue:                     workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "longhorn-uninstall"),
+	}
+
+	volumeInformer.Informer().AddEventHandler(c.controlleeHandler())
+	engineInformer.Informer().AddEventHandler(c.controlleeHandler())
+	replicaInformer.Informer().AddEventHandler(c.controlleeHandler())
+	engineImageInformer.Informer().AddEventHandler(c.controlleeHandler())
+	nodeInformer.Informer().AddEventHandler(c.controlleeHandler())
+	daemonSetInformer.Informer().AddEventHandler(c.controlleeHandler())
+	return c
+}
+
+func (c *UninstallController) controlleeHandler() cache.ResourceEventHandler {
+	return cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { c.enqueueControlleeChange() },
+		UpdateFunc: func(old, cur interface{}) { c.enqueueControlleeChange() },
+		DeleteFunc: func(obj interface{}) { c.enqueueControlleeChange() },
+	}
+}
+
+func (c *UninstallController) enqueueControlleeChange() {
+	c.queue.AddRateLimited("uninstall")
+}
+
+func (c *UninstallController) Run() {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	logrus.Infof("Uninstalling...")
+
+	if !controller.WaitForCacheSync("longhorn uninstall", c.stopCh,
+		c.volumeInformerSynced, c.engineImageInformerSynced,
+		c.nodeInformerSynced, c.daemonSetInformerSynced) {
+
+		logrus.Fatalf("Failed to sync informers")
+	}
+
+	go wait.Until(c.worker, time.Second, c.stopCh)
+	<-c.stopCh
+	logrus.Infof("Uninstallation complete")
+}
+
+func (c *UninstallController) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *UninstallController) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	err := c.uninstall()
+	c.handleErr(err, key)
+
+	return true
+}
+
+func (c *UninstallController) handleErr(err error, key interface{}) {
+	if err == nil {
+		c.queue.Forget(key)
+		return
+	}
+
+	logrus.Warnf("Error: %v", err)
+	c.queue.AddRateLimited(key)
+}
+
+func (c *UninstallController) uninstall() error {
+	if c.firstRun {
+		if err := c.checkReadyShutdown(); err != nil {
+			return err
+		}
+		c.firstRun = false
+	}
+
+	if done, err := c.deleteCRDs(); err != nil {
+		return err
+	} else if !done {
+		return nil
+	}
+
+	if done, err := c.deleteManager(); err != nil {
+		return err
+	} else if !done {
+		return nil
+	}
+
+	go func() {
+		for {
+			select {
+			case c.stopCh <- struct{}{}:
```

Comment:
Don't need nonblocking here. If there is something in stopCh, the controller should be already stopped.